<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Shubham Dutta">
<meta name="dcterms.date" content="2023-03-17">

<title>Shubham Dutta - Real time qPCR data analysis using R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<link href="../../site_libs/quarto-contrib/academicons-1.9.2/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/academicons-1.9.2/size.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Shubham Dutta</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../projects.html"> 
<span class="menu-text">Projects</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../blogs.html"> 
<span class="menu-text">Blogs</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../experience.html"> 
<span class="menu-text">Experience</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../skills.html"> 
<span class="menu-text">Skills</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../publications.html"> 
<span class="menu-text">Publications</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="mailto:shubhamdutta26@gmail.com"> <i class="bi bi-envelope" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://scholar.google.com/citations?user=okp_AlgAAAAJ&amp;hl=en"> 
<span class="menu-text"><i class="ai  ai-google-scholar" title="" style="color:"></i></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/shubhamdutta26/"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://orcid.org/0000-0001-8484-0717"> 
<span class="menu-text"><i class="ai  ai-orcid" title="" style="color:"></i></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/shubhamdutta26"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://stackoverflow.com/users/17288160/shubham"> 
<span class="menu-text"><i class="ai  ai-stackoverflow" title="" style="color:"></i></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Real time qPCR data analysis using R</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">qpcr</div>
                <div class="quarto-category">sybr green</div>
                <div class="quarto-category">r</div>
                <div class="quarto-category">tidyverse</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Shubham Dutta </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">March 17, 2023</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level3">
<h3 class="anchored" data-anchor-id="introduction">Introduction</h3>
<p>I will be using the <a href="https://www.tidyverse.org/">Tidyverse</a> package in R to analyse and plot a simple RT-qPCR experiment data. In this experiment, I have ectopically-expressed two proteins namely NKG2A and NKG2C (along with their binding partner CD94) in HEK293 cells. I want to assess the expression status of these genes in the HEK293 cells relative to GAPDH using the <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> method. The experiment was run in a <a href="https://www.bio-rad.com/en-us/product/c1000-touch-thermal-cycler?ID=LGTW9415">C1000 Touch Thermal Cycler</a> instrument. The csv file can be found <a href="https://github.com/shubhamdutta26/qPCR-data-analysis-R/blob/main/qPCR_HEKs_OE_NKG2A-C_human.csv">here</a>.</p>
</section>
<section id="the-deltadeltact-method-for-gene-expression-analysis" class="level3">
<h3 class="anchored" data-anchor-id="the-deltadeltact-method-for-gene-expression-analysis">The <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> method for gene expression analysis</h3>
<p>The <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> method (also known as the 2<sup>–∆∆C<sub>T</sub></sup> method) is a simple formula used to calculate the relative fold gene expression of samples when performing RT-qPCR. The method was developed by Kenneth Livak and Thomas Schmittgen in 2001 and is widely used (<a href="https://doi.org/10.1006/meth.2001.1262">Livak &amp; Schmittgen, 2001</a>).</p>
<p>C<sub>T</sub> (or C<sub>Q</sub>) stands for the Cycle Threshold (or Cycle Quantification) of a sample. This is given after the qPCR reaction by the software. Simply put, it is the cycle number where the fluorescence generated by the PCR product is distinguishable from the background fluorescence signal.</p>
<p><em>The symbol</em> <span class="math inline">\(\Delta\)</span> (Delta) is a mathematical term used to describe the difference between two values.</p>
<p>Let’s see how we can calculate the <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub>:</p>
<blockquote class="blockquote">
<p><span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> = <span class="math inline">\(\Delta\)</span>C<sub>T</sub> (treated sample) – <span class="math inline">\(\Delta\)</span>C<sub>T</sub> (untreated sample)</p>
</blockquote>
<p>Essentially, <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> is the difference between the <span class="math inline">\(\Delta\)</span>C<sub>T</sub> values of the treated/experimental sample and the untreated/control sample. But what does <span class="math inline">\(\Delta\)</span>C<sub>T</sub> refer to?</p>
<blockquote class="blockquote">
<p><span class="math inline">\(\Delta\)</span>C<sub>T</sub> = C<sub>T</sub> (gene of interest) – C<sub>T</sub> (housekeeping gene)</p>
</blockquote>
<p>Basically, <span class="math inline">\(\Delta\)</span>C<sub>T</sub> is the difference in <span class="math inline">\(\Delta\)</span>C<sub>T</sub> values for your gene of interest and your housekeeping gene for a given sample. This is to essentially normalise the gene of interest to a gene which is not affected by your experiment.</p>
<p>Finally, to work out the fold gene expression (F<sub>C</sub>) we need to do 2 to the power of negative <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub>:</p>
<blockquote class="blockquote">
<p>F<sub>C</sub> = 2<sup>–</sup><span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub></p>
</blockquote>
<p>Let us move to the fun part now!</p>
</section>
<section id="data-analysis" class="level3">
<h3 class="anchored" data-anchor-id="data-analysis">Data Analysis</h3>
<section id="install-and-load-tidyverse-package" class="level4">
<h4 class="anchored" data-anchor-id="install-and-load-tidyverse-package">Install and load Tidyverse package</h4>
<p>We will use the <a href="https://www.tidyverse.org/">Tidyverse</a> package for the data analysis.</p>
<p>Load the <code>tidyverse</code> and <code>readr</code> packages into R Studio environment:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(janitor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="importing-the-data" class="level4">
<h4 class="anchored" data-anchor-id="importing-the-data">Importing the data</h4>
<p>Now we will import the qPCR data (a csv document in this case) into the R Studio environment using the <code>read.csv</code> function and store it as a dataframe (df) called <strong>qpcr</strong>.<br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>qpcr <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"qPCR_HEKs_OE_NKG2A-C_human.csv"</span>) <span class="sc">%&gt;%</span> <span class="fu">clean_names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 60 Columns: 15
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (6): Well, Fluor, Target, Content, Sample, Biological Set Name
dbl (8): Cq, Cq Mean, Cq Std. Dev, Starting Quantity (SQ), Log Starting Quan...
lgl (1): Well Note

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tibble</span>(qpcr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 60 × 15
   well  fluor target content sample   biological_set_name    cq cq_mean
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt;   &lt;dbl&gt;
 1 A01   SYBR  GAPDH  Unkn    Sample 1 empty vector         14.5    14.5
 2 A02   SYBR  NKG2A  Unkn    Sample 1 empty vector         20.0    20.0
 3 A03   SYBR  NKG2C  Unkn    Sample 1 empty vector         17.5    17.5
 4 A04   SYBR  CD94   Unkn    Sample 1 empty vector         19.8    19.8
 5 A05   SYBR  GAPDH  Unkn    Sample 1 pcDNA-NKG2A          14.5    14.5
 6 A06   SYBR  NKG2A  Unkn    Sample 1 pcDNA-NKG2A          14.3    14.3
 7 A07   SYBR  NKG2C  Unkn    Sample 1 pcDNA-NKG2A          19.2    19.2
 8 A08   SYBR  CD94   Unkn    Sample 1 pcDNA-NKG2A          14.4    14.4
 9 A09   SYBR  GAPDH  Unkn    Sample 1 pcDNA-NKG2C          14.5    14.5
10 A10   SYBR  NKG2A  Unkn    Sample 1 pcDNA-NKG2C          22.3    22.3
# ℹ 50 more rows
# ℹ 7 more variables: cq_std_dev &lt;dbl&gt;, starting_quantity_sq &lt;dbl&gt;,
#   log_starting_quantity &lt;dbl&gt;, sq_mean &lt;dbl&gt;, sq_std_dev &lt;dbl&gt;,
#   set_point &lt;dbl&gt;, well_note &lt;lgl&gt;</code></pre>
</div>
</div>
<p>There are 15 columns (variables) in our data but we only need a some them for our data analysis. Therefore, using the <code>select</code> function we will extract the columns we need and store them in a new df called <strong>tidyData</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>tidyData <span class="ot">&lt;-</span> qpcr <span class="sc">%&gt;%</span> <span class="fu">select</span>(target<span class="sc">:</span>cq, <span class="sc">-</span>content)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(tidyData)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 60
Columns: 4
$ target              &lt;chr&gt; "GAPDH", "NKG2A", "NKG2C", "CD94", "GAPDH", "NKG2A…
$ sample              &lt;chr&gt; "Sample 1", "Sample 1", "Sample 1", "Sample 1", "S…
$ biological_set_name &lt;chr&gt; "empty vector", "empty vector", "empty vector", "e…
$ cq                  &lt;dbl&gt; 14.45711, 20.02642, 17.51733, 19.79295, 14.48283, …</code></pre>
</div>
</div>
<p>Lets look at the 4 columns now:</p>
<ul>
<li><strong>target</strong>: This variable stores the different target genes which in this case are GAPDH, NKG2A, NKG2B, and CD94.</li>
<li><strong>sample</strong>: These are different biological replicates. There are 5 replicates. There are no technical replicates for this experiment.</li>
<li><strong>biological_set_name</strong>: This variable stores the three different conditions. HEK cells are transfected with empty vector (empty vector), NKG2A/ CD94 (pcDNA-NKG2A), and NKG2C/ CD94 (pcDNA-NKG2C).</li>
<li>**cq*: The C<sub>q</sub> (or C<sub>T</sub>) value for each condition. From here on till the end I will be using C<sub>T</sub> and C<sub>Q</sub> interchangeably.</li>
</ul>
<p><br></p>
</section>
<section id="calculate-deltact" class="level4">
<h4 class="anchored" data-anchor-id="calculate-deltact">Calculate <span class="math inline">\(\Delta\)</span>C<sub>T</sub></h4>
<p>The <span class="math inline">\(\Delta\)</span>C<sub>T</sub> is the differnce between the C<sub>T</sub> of the housekeeping gene and the target gene. To calculate <span class="math inline">\(\Delta\)</span>C<sub>T</sub> we need to determine the average C<sub>T</sub> of GAPDH for three conditions (empty vector, pcDNA-NKG2A, and pcDNA-NKG2C).</p>
<p>We will use the <code>filter</code> function to extract and store the C<sub>T</sub> values of GAPDH in a new df called <strong>gapdh_only</strong>. Then we will use the <code>group_by</code>, <code>summarise</code>, and <code>mean()</code> functions to calculate mean C<sub>T</sub>s for the three conditions and store the values in another df called <strong>avg_gapdh</strong>.</p>
<p><code>%&gt;%</code> is called the pipe operator. This special operational function available under the <a href="https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html">magrittr</a> and <a href="https://dplyr.tidyverse.org/">dplyr</a> package (basically developed under magrittr), which allows us to pass the result of one function/argument to the other one in sequence.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gapdh_only <span class="ot">&lt;-</span> tidyData <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(target <span class="sc">==</span> <span class="st">"GAPDH"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>avg_gapdh <span class="ot">&lt;-</span> gapdh_only <span class="sc">%&gt;%</span> </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(biological_set_name) <span class="sc">%&gt;%</span> </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">ref_cq =</span> <span class="fu">mean</span>(cq))</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>avg_gapdh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  biological_set_name ref_cq
  &lt;chr&gt;                &lt;dbl&gt;
1 empty vector          14.5
2 pcDNA-NKG2A           14.5
3 pcDNA-NKG2C           14.5</code></pre>
</div>
</div>
<p>We have stored the mean GAPDH C<sub>T</sub> value as <em>ref_cq</em>. We will subtract the <em>ref_cq</em> from the <em>cq</em> of the other target genes. Let’s take <em>NKG2A</em> as an example and see a step-by-step process:</p>
<p>First, we will extract the NKG2A target gene data for all the conditions and store it in <em>nkg2a_only</em> df.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>nkg2a_only <span class="ot">&lt;-</span> tidyData <span class="sc">%&gt;%</span> <span class="fu">filter</span>(target <span class="sc">==</span> <span class="st">"NKG2A"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nkg2a_only)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  target sample   biological_set_name    cq
  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt;
1 NKG2A  Sample 1 empty vector         20.0
2 NKG2A  Sample 1 pcDNA-NKG2A          14.3
3 NKG2A  Sample 1 pcDNA-NKG2C          22.3
4 NKG2A  Sample 2 empty vector         20.0
5 NKG2A  Sample 2 pcDNA-NKG2A          14.4
6 NKG2A  Sample 2 pcDNA-NKG2C          22.3</code></pre>
</div>
</div>
<p>Second, we will join the <em>avg_gapdh</em> &amp; <em>nkg2a_only</em> dfs using <code>left_join</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>nkg2a_gapdh <span class="ot">&lt;-</span> <span class="fu">left_join</span>(nkg2a_only, avg_gapdh, <span class="at">by =</span> <span class="st">"biological_set_name"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nkg2a_gapdh)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  target sample   biological_set_name    cq ref_cq
  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt;  &lt;dbl&gt;
1 NKG2A  Sample 1 empty vector         20.0   14.5
2 NKG2A  Sample 1 pcDNA-NKG2A          14.3   14.5
3 NKG2A  Sample 1 pcDNA-NKG2C          22.3   14.5
4 NKG2A  Sample 2 empty vector         20.0   14.5
5 NKG2A  Sample 2 pcDNA-NKG2A          14.4   14.5
6 NKG2A  Sample 2 pcDNA-NKG2C          22.3   14.5</code></pre>
</div>
</div>
<p>Finally, we will calculate the <span class="math inline">\(\Delta\)</span>C<sub>T</sub> for the NKG2A dataset and store the values in <em>nkg2a_gapdh</em> df. The <code>mutate</code> function basically calculates <em>del_cq</em> values and adds a new column (variable) to the existing df.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>nkg2a_gapdh <span class="ot">&lt;-</span> nkg2a_gapdh <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">del_cq =</span> cq <span class="sc">-</span> ref_cq)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nkg2a_gapdh)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 6
  target sample   biological_set_name    cq ref_cq  del_cq
  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;
1 NKG2A  Sample 1 empty vector         20.0   14.5  5.57  
2 NKG2A  Sample 1 pcDNA-NKG2A          14.3   14.5 -0.135 
3 NKG2A  Sample 1 pcDNA-NKG2C          22.3   14.5  7.85  
4 NKG2A  Sample 2 empty vector         20.0   14.5  5.59  
5 NKG2A  Sample 2 pcDNA-NKG2A          14.4   14.5 -0.0720
6 NKG2A  Sample 2 pcDNA-NKG2C          22.3   14.5  7.88  </code></pre>
</div>
</div>
<p>Please feel free to use the dataset to calculate the <span class="math inline">\(\Delta\)</span>C<sub>T</sub> for NKG2C and CD94 genes.</p>
</section>
<section id="calculate-deltadeltact" class="level4">
<h4 class="anchored" data-anchor-id="calculate-deltadeltact">Calculate <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub></h4>
<p>The <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> is the difference between the <span class="math inline">\(\Delta\)</span>C<sub>T</sub> values of experimental sample and the control sample.</p>
<p>First, we calculate mean C<sub>T</sub> for “empty vector” from <em>nkg2a_gapdh</em> df and store it in the <strong>nkg2a_empty</strong> variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>nkg2a_empty <span class="ot">&lt;-</span> nkg2a_gapdh <span class="sc">%&gt;%</span> <span class="fu">filter</span>(biological_set_name <span class="sc">==</span> <span class="st">"empty vector"</span>) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">mean_empty =</span> <span class="fu">mean</span>(del_cq)) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(mean_empty)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>nkg2a_empty</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.592248</code></pre>
</div>
</div>
<p>Second, we calculate the <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> values and append the values to <em>nkg2a_gapdh</em> df.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>nkg2a_gapdh <span class="ot">&lt;-</span> nkg2a_gapdh <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">del_del_cq =</span> del_cq <span class="sc">-</span> nkg2a_empty)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nkg2a_gapdh)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
  target sample   biological_set_name    cq ref_cq  del_cq del_del_cq
  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;
1 NKG2A  Sample 1 empty vector         20.0   14.5  5.57     -0.0191 
2 NKG2A  Sample 1 pcDNA-NKG2A          14.3   14.5 -0.135    -5.73   
3 NKG2A  Sample 1 pcDNA-NKG2C          22.3   14.5  7.85      2.26   
4 NKG2A  Sample 2 empty vector         20.0   14.5  5.59     -0.00487
5 NKG2A  Sample 2 pcDNA-NKG2A          14.4   14.5 -0.0720   -5.66   
6 NKG2A  Sample 2 pcDNA-NKG2C          22.3   14.5  7.88      2.29   </code></pre>
</div>
</div>
<p>Next, we can calculate the fold change gene expression for each condition using the F<sub>C</sub> formula above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>nkg2a_geneExp <span class="ot">&lt;-</span> nkg2a_gapdh <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">geneExp =</span> <span class="dv">2</span><span class="sc">^</span>(del_del_cq))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nkg2a_geneExp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  target sample   biological_set_name    cq ref_cq  del_cq del_del_cq geneExp
  &lt;chr&gt;  &lt;chr&gt;    &lt;chr&gt;               &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;      &lt;dbl&gt;   &lt;dbl&gt;
1 NKG2A  Sample 1 empty vector         20.0   14.5  5.57     -0.0191   0.987 
2 NKG2A  Sample 1 pcDNA-NKG2A          14.3   14.5 -0.135    -5.73     0.0189
3 NKG2A  Sample 1 pcDNA-NKG2C          22.3   14.5  7.85      2.26     4.78  
4 NKG2A  Sample 2 empty vector         20.0   14.5  5.59     -0.00487  0.997 
5 NKG2A  Sample 2 pcDNA-NKG2A          14.4   14.5 -0.0720   -5.66     0.0197
6 NKG2A  Sample 2 pcDNA-NKG2C          22.3   14.5  7.88      2.29     4.89  </code></pre>
</div>
</div>
<p>Finally, let us calculate statistical parameters using <code>mean()</code> and <code>sd()</code> for the data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>nkg2a_geneExp_final <span class="ot">&lt;-</span> nkg2a_geneExp <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(biological_set_name) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">mean_geneExp =</span> <span class="fu">mean</span>(geneExp), <span class="at">sd_geneExp =</span> <span class="fu">sd</span>(geneExp)) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Target =</span> <span class="fu">c</span>(<span class="st">"NKG2A"</span>, <span class="st">"NKG2A"</span>, <span class="st">"NKG2A"</span>))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(nkg2a_geneExp_final)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  biological_set_name mean_geneExp sd_geneExp Target
  &lt;chr&gt;                      &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt; 
1 empty vector              1.00     0.0211   NKG2A 
2 pcDNA-NKG2A               0.0196   0.000423 NKG2A 
3 pcDNA-NKG2C               4.75     0.127    NKG2A </code></pre>
</div>
</div>
<p>Please feel free to use the dataset to calculate the <span class="math inline">\(\Delta\)</span><span class="math inline">\(\Delta\)</span>C<sub>T</sub> for NKG2C and CD94 genes.</p>
<p>Lets move on to the visualization of our data.</p>
</section>
</section>
<section id="data-visualization" class="level3">
<h3 class="anchored" data-anchor-id="data-visualization">Data Visualization</h3>
<p>Before plotting the data let us make a single df. We will use <code>full_join</code> to accomplish this and make a new df called <strong>all</strong> which contains the analysis for individual targets namely NKG2C (<strong>nkg2c_geneExp_final</strong>) and CD94 (<strong>cd94_geneExp_final</strong>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>all <span class="ot">&lt;-</span> <span class="fu">full_join</span>(nkg2a_geneExp_final, nkg2c_geneExp_final)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>all <span class="ot">&lt;-</span> <span class="fu">full_join</span>(all, cd94_geneExp_final)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>all</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 9 × 5
  biological_set_name mean_geneExp sd_geneExp Target target
  &lt;chr&gt;                      &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; 
1 empty vector              1.00     0.0211   NKG2A  &lt;NA&gt;  
2 pcDNA-NKG2A               0.0196   0.000423 NKG2A  &lt;NA&gt;  
3 pcDNA-NKG2C               4.75     0.127    NKG2A  &lt;NA&gt;  
4 empty vector              1.00     0.0727   &lt;NA&gt;   NKG2C 
5 pcDNA-NKG2A               0.266    0.0564   &lt;NA&gt;   NKG2C 
6 pcDNA-NKG2C               8.55     0.0526   &lt;NA&gt;   NKG2C 
7 empty vector              1.02     0.241    &lt;NA&gt;   CD94  
8 pcDNA-NKG2A              35.2      0.179    &lt;NA&gt;   CD94  
9 pcDNA-NKG2C              34.8      0.109    &lt;NA&gt;   CD94  </code></pre>
</div>
</div>
<p>Let us now use <code>ggplot</code> package to plot a bar plot for the RT-qPCR data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(all, <span class="fu">aes</span>(<span class="at">x =</span> biological_set_name, <span class="at">y =</span> mean_geneExp, <span class="at">fill =</span> target)) <span class="sc">+</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">width =</span> <span class="fl">0.5</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.6</span>), <span class="at">color =</span> <span class="st">"#434343"</span>) <span class="sc">+</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> mean_geneExp <span class="sc">-</span> sd_geneExp, <span class="at">ymax =</span> mean_geneExp <span class="sc">+</span> sd_geneExp), <span class="at">width=</span>.<span class="dv">2</span>, <span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="fl">0.6</span>)) <span class="sc">+</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()<span class="sc">+</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Human NKG2A/C-CD94 overexpression status in HEK293 cells"</span>,</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"2021.11.23 Shubham Dutta"</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Transfection (Human genes)"</span>,</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Expression relative to GAPDH"</span>,</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"Target gene"</span>,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">caption =</span> <span class="cn">NULL</span>)<span class="sc">+</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">13</span>),</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>, <span class="at">color =</span> <span class="st">"#434343"</span>),</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">13</span>),</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">10</span>),</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>),</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.8</span>),</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>())<span class="sc">+</span></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">"#d55e00"</span>, <span class="st">"#0072b2"</span>, <span class="st">"#009e73"</span>)) <span class="sc">+</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">expand =</span> <span class="fu">expansion</span>(<span class="at">mult =</span> <span class="fu">c</span>(<span class="dv">0</span>, .<span class="dv">1</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: A numeric `legend.position` argument in `theme()` was deprecated in ggplot2
3.5.0.
ℹ Please use the `legend.position.inside` argument of `theme()` instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="references" class="level3">
<h3 class="anchored" data-anchor-id="references">References</h3>
<ol type="1">
<li>Kenneth J. Livak, Thomas D. Schmittgen, Analysis of Relative Gene Expression Data Using Real-Time Quantitative PCR and the 2−ΔΔCT Method, Methods, Volume 25, Issue 4, 2001, Pages 402-408, ISSN 1046-2023, https://doi.org/10.1006/meth.2001.1262.</li>
</ol>
<p><br><br></p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>